src/util/src/tcache.c

```diff
@@ -19,688 +19,581 @@
-         pCacheObj->name, pNode->key, pNode->data, pNode->size, size - 1, pCacheObj->totalSize);
+  pthread_mutex_lock(&tsCacheMutex);
 
-  if (pCacheObj->freeFp) {
-    pCacheObj->freeFp(pNode->data);
+  int i;
+  for (i=0; i < TAOS_CACHE_MAX_OBJS; ++i) {
+    tsCacheNextId = (tsCacheNextId + 1) % TAOS_CACHE_MAX_OBJS;
+    if (cacheObjList[tsCacheNextId].state == TAOS_CACHE_STATE_EMPTY) break;
   }
 
-  free(pNode);
+  if (i >= TAOS_CACHE_MAX_OBJS) {
+    terrno = TSDB_CODE_CACHE_TOO_MANY; 
+    uError("cache:%s too many cache objs", name);
+    free(p1);
+    free(p2);
+  } else {
+    cacheId = tsCacheNextId;
+    SCacheObj *pCacheObj = cacheObjList + cacheId;
+
+    strncpy(pCacheObj->name, name, sizeof(pCacheObj->name)-1);
+    pCacheObj->size = size;
+    pCacheObj->refreshTime = refreshTimeInSeconds * 1000;
+    pCacheObj->extendLifespan = extendLifespan;
+    pCacheObj->cacheId = cacheId;  
+    pCacheObj->hashFp = taosGetDefaultHashFunction(keyType);
+    pCacheObj->freeFp = fn;
+    pCacheObj->nodeList = p1;
+    pCacheObj->nlock = p2;
+    pCacheObj->state = TAOS_CACHE_STATE_ACTIVE;
+ 
+    // start timer
+    if (tsCacheNum == 0) 
+      tsCacheTmrCtrl = taosTmrInit( TAOS_CACHE_MAX_OBJS+1, 1000, 1000000, "cache");
+    pCacheObj->pTimer = taosTmrStart(taosCacheProcessTimer, pCacheObj->refreshTime, pCacheObj, tsCacheTmrCtrl);  
+    cacheId = tsCacheNextId;
+    tsCacheNum++;
+  }
+
+
+  pthread_mutex_unlock(&tsCacheMutex);
+
+  uTrace("cache:%s is initialized, cacheId:%d total cache:%d", name, cacheId, tsCacheNum);
+  return cacheId;
 }
 
-static FORCE_INLINE STrashElem* doRemoveElemInTrashcan(SCacheObj* pCacheObj, STrashElem *pElem) {
-  if (pElem->pData->signature != (uint64_t) pElem->pData) {
-    uWarn("key:sig:0x%" PRIx64 " %p data has been released, ignore", pElem->pData->signature, pElem->pData);
+void *taosCachePut(int32_t cacheId, const void *key, size_t keyLen, const void *pData, size_t dataSize, int durationMS) {
+  if (cacheId < 0 || cacheId >= TAOS_CACHE_MAX_OBJS) {
+    terrno = TSDB_CODE_CACHE_INVALID_ID;
     return NULL;
   }
 
-  STrashElem* next = pElem->next;
-
-  pCacheObj->numOfElemsInTrash--;
-  if (pElem->prev) {
-    pElem->prev->next = pElem->next;
-  } else { // pnode is the header, update header
-    pCacheObj->pTrash = pElem->next;
+  // create a new node
+  size_t size = dataSize + sizeof(SCacheNode) + keyLen;
+  SCacheNode *pNewNode = calloc(1, size);
+  if (pNewNode == NULL) {
+    terrno = TSDB_CODE_COM_OUT_OF_MEMORY;
+    return NULL;
+  } 
+
+  SCacheObj *pCacheObj = cacheObjList + cacheId;
+  if (pCacheObj->count * TAOS_CACHE_LOAD_FACTOR > pCacheObj->size && pCacheObj->iterNum <= 0) {
+    taosWLockLatch(&pCacheObj->olock);
+    taosCacheResize(pCacheObj);
+    taosWUnLockLatch(&pCacheObj->olock);
+  }
+ 
+  uint32_t hash = (*pCacheObj->hashFp)(key, (uint32_t)keyLen) % pCacheObj->size;
+  taosRLockLatch(&pCacheObj->olock);
+  if (pCacheObj->state != TAOS_CACHE_STATE_ACTIVE) {
```

Comment:
the check of `pCacheObj->state` is protected by `pCacheObj->olock` here, but by `tsCacheMutex` in `taosCacheCleanup`, which may be a race condition